libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt

```diff
@@ -0,0 +1,112 @@
+package com.mapbox.navigation.ui.voice.api
+
+import android.content.Context
+import com.mapbox.api.directions.v5.models.VoiceInstructions
+import com.mapbox.navigation.ui.base.model.voice.Announcement
+import com.mapbox.navigation.ui.voice.VoiceAction
+import com.mapbox.navigation.ui.voice.VoiceProcessor
+import com.mapbox.navigation.ui.voice.VoiceResult
+import com.mapbox.navigation.ui.voice.model.VoiceState
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import okhttp3.ResponseBody
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.io.File
+
+/**
+ * Implementation of [VoiceApi] allowing you to retrieve voice instructions.
+ * @property context Context
+ * @property accessToken String
+ * @property language String
+ */
+internal class MapboxVoiceApi(
+    private val context: Context,
+    private val accessToken: String,
+    private val language: String
+) : VoiceApi {
+
+    /**
+     * Given [VoiceInstructions] the method returns a [File] wrapped inside [VoiceCallback]
+     * @param voiceInstruction VoiceInstructions object representing [VoiceInstructions]
+     * @param callback VoiceCallback contains [VoiceState.VoiceFile]
+     */
+    override fun retrieveVoiceFile(voiceInstruction: VoiceInstructions, callback: VoiceCallback) {
+        val action = VoiceAction.PrepareVoiceRequest(voiceInstruction)
+        when (val result = VoiceProcessor.process(action)) {
+            is VoiceResult.VoiceRequest -> {
+                val mapboxSpeech = result.request
+                    .accessToken(accessToken)
+                    .language(language)
+                    .build()
+                mapboxSpeech.enqueueCall(object : Callback<ResponseBody> {
+                    override fun onResponse(
+                        call: Call<ResponseBody>,
+                        response: Response<ResponseBody>
+                    ) {
+                        val voiceAction = VoiceAction.ProcessVoiceResponse(response)
+                        when (val res = VoiceProcessor.process(voiceAction)) {
+                            is VoiceResult.Voice.Success -> {
+                                ThreadController.getMainScopeAndRootJob().scope.launch {
```

Comment:
idea: assign each VoiceInstructions a Job, cancel would cancel the job

Reply:
It's not that easy as that'd only cover the case for cancelling storing the file into disk but neither the rest of the callbacks nor the `mapboxSpeech.enqueueCall`.